### PR TITLE
resolve #8

### LINKED ISSUE #8: Add missing compiler flags.
Currently, we use default compiler flags (except for using c++20). 
We must do some analysis, what flags should we use by default in each functional cluster, and add them.

One of them is `-Wall`.  For now, we consider only `GCC 10.0+`

### PR DESCRIPTION
**Pull request by pz0**
I added few gcc flags to meson.build file. Do you have any suggestions about flags?

### CODE REVIEW

**Comment by satletio1** [comment]
@UmlautSoftwareDevelopmentAccount/adaptive-platform 

What additional flags can be used here?

File: meson.build [hunk 1]
```diff
@@ -4,6 +4,9 @@ project('ap-functional-cluster-template', 'cpp',
     default_options : ['cpp_std=c++2a']
 )
 
+# additional compilation arguments
+add_global_arguments('-Wall', '-Wextra', '-O2', language : 'cpp')
```
Comment (bburda) [inline meson.build:8]: Well, there are many more flags that we can consider, and I think we should discuss it deeper and finish this task with the ADR on the main repository.

Flags to consider: 
`-Wall`
`-Wextra`
`-Werror`
`-Wshadow`
`-Wdouble-promotion`
`-Wconversion`
`-Wpendantic`
`-pie`
`-flto`
`-O2`

What you think @insooth @slci @hliberacki?

Comment (slci) [reply meson.build:8]: I think -Ox and lto flags should not be hard-coded, those are under control of Meson build

Comment (slci) [reply meson.build:8]: Regarding other flags we could make use Jason's list from his CMake cpp_starter_project: https://github.com/lefticus/cpp_starter_project/blob/b380f3d22b9426575f1b7e9135908a94c00a18a9/cmake/CompilerWarnings.cmake#L43
and port it to Meson. What do you think?

Comment (bburda) [reply meson.build:8]: Sounds good. And what you think about `-Wpedantic` and `-Werror`?

Comment (slci) [reply meson.build:8]: I'd use -Wpedantic as it helps write portable code and not enable it only in specific, well documented cases, where we'd like to use some extensions to the language.
-Werror should also be enabled by default in my opinion.

**Comment by bburda** [comment]
@pz0 any update on this?

File: meson.build (hunk 1 above)
Comment (slci) [reply meson.build:8]: Please check #17 for default warnings and align.

Comment (pz0) [reply meson.build]: I added build-in options from #17 and additional warnings from: https://github.com/lefticus/cpp_starter_project/blob/b380f3d22b9426575f1b7e9135908a94c00a18a9/cmake/CompilerWarnings.cmake#L43

File: meson.build [hunk 2]
```diff
@@ -2,14 +2,38 @@ project('ap-functional-cluster-template', 'cpp',
...
+		'warning_level=3',
+		'werror=true',
+		'optimization=2',
+		'pkg_config_path=.',
+		'b_coverage=true'
     ]
 )
 
+# additional compilation arguments
+add_global_arguments(
```
Comment (bburda) [inline meson.build:15]: Meson has some of these options already built-in and should be set differently.
Please have a look at: https://mesonbuild.com/Builtin-options.html 
I'm not seeing `-Wshadow`, why?

@slci Please also look at this change

**Comment by slci** [comment]
@bburda 
The default options are good IMO. `'warning_level=3'` and `'werror=true'` do the job. But as you said, `-Wshadow` can be added.

**Comment by bburda** [comment]
> 
> 
> @bburda
> The default options are good IMO. `'warning_level=3'` and `'werror=true'` do the job. But as you said, `-Wshadow` can be added.

Are you sure it is enough?
- warning_level 1 adds -Wall,
- warning_level 2 adds warning_level 1 + -Wextra,
- warning_level 3 adds warning_level 2 + -Wpedantic

File: meson.build [hunk 3]
```diff
@@ -2,14 +2,38 @@ project('ap-functional-cluster-template', 'cpp',
...
     default_options : [
-        'cpp_std=c++2a',
-        'warning_level=3',
-        'werror=true',
-        'pkg_config_path=.',
-        'b_coverage=true'
+		'cpp_std=c++2a',
+		'warning_level=3',
+		'werror=true',
+		'optimization=2',
```
Comment (bburda) [inline meson.build:8]: We should use `buildtype` instead of `optimization` in my opinion.

**Comment by slci** [comment]
> > @bburda
> > The default options are good IMO. `'warning_level=3'` and `'werror=true'` do the job. But as you said, `-Wshadow` can be added.
> 
> Are you sure it is enough?
> 
>     * warning_level 1 adds -Wall,
> 
>     * warning_level 2 adds warning_level 1 + -Wextra,
> 
>     * warning_level 3 adds warning_level 2 + -Wpedantic

My point is that default `warning_level 3` together with the additional flags added here in this change are fine.

**Comment by bburda** [comment]
> 
> 
> > > @bburda
> > > The default options are good IMO. `'warning_level=3'` and `'werror=true'` do the job. But as you said, `-Wshadow` can be added.
> > 
> > 
> > Are you sure it is enough?
> > ```
> > * warning_level 1 adds -Wall,
> > 
> > * warning_level 2 adds warning_level 1 + -Wextra,
> > 
> > * warning_level 3 adds warning_level 2 + -Wpedantic
> > ```
> 
> My point is that default `warning_level 3` together with the additional flags added here in this change are fine.
> The toolchain I have gives the following flags on compilation" `-Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -Wpedantic -Werror -std=c++2a -g -Wnon-virtual-dtor -Wold-style-cast -Wcast-align -Wunused -Woverloaded-virtual -Wconversion -Wsign-conversion -Wnull-dereference -Wdouble-promotion -Wformat=2 -Wmisleading-indentation -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wuseless-cast -pie -fpie -flto -fPIC` so only `-Wshadow` is missing for me.

Right! Just some of them should be added through Meson built-in options
https://mesonbuild.com/Builtin-options.html#base-options (lto and pie for sure)

**Comment by slci** [comment]
> > > > @bburda
> > > > The default options are good IMO. `'warning_level=3'` and `'werror=true'` do the job. But as you said, `-Wshadow` can be added.
> > > 
> > > 
> > > Are you sure it is enough?
> > > ```
> > > * warning_level 1 adds -Wall,
> > > 
> > > * warning_level 2 adds warning_level 1 + -Wextra,
> > > 
> > > * warning_level 3 adds warning_level 2 + -Wpedantic
> > > ```
> > 
> > 
> > My point is that default `warning_level 3` together with the additional flags added here in this change are fine.
> > The toolchain I have gives the following flags on compilation" `-Wall -Winvalid-pch -Wnon-virtual-dtor -Wextra -Wpedantic -Werror -std=c++2a -g -Wnon-virtual-dtor -Wold-style-cast -Wcast-align -Wunused -Woverloaded-virtual -Wconversion -Wsign-conversion -Wnull-dereference -Wdouble-promotion -Wformat=2 -Wmisleading-indentation -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wuseless-cast -pie -fpie -flto -fPIC` so only `-Wshadow` is missing for me.
> 
> Right! Just some of them should be added through Meson built-in options
> https://mesonbuild.com/Builtin-options.html#base-options (lto and pie for sure)

Ah, ok, I see. Sure, I didn't notice Meson already has `b_lto` and `b_pie`.

File: meson.build (hunk 2 above)
Comment (pz0) [reply meson.build:15]: done

File: meson.build (hunk 3 above)
Comment (pz0) [reply meson.build]: done

File: meson.build [hunk 4]
```diff
@@ -2,14 +2,39 @@ project('ap-functional-cluster-template', 'cpp',
...
-        'warning_level=3',
-        'werror=true',
-        'pkg_config_path=.',
-        'b_coverage=true'
+		'cpp_std=c++2a',
+		'warning_level=3',
+		'werror=true',
+		'buildtype=debugoptimized',
+		'pkg_config_path=.',
+		'b_coverage=true'
```
Comment (bburda) [inline meson.build:10]: missing `,`

**Pull request merged**